allennlp/tests/modules/drop_connect_test.py

```diff
@@ -90,14 +121,38 @@ def test_parameters_are_leaf_tensors(self):
         assert all(parameter.is_leaf for parameter in weight_dropped_linear.parameters())
 
         # Case 3: After forward
-        input_tensor = torch.ones(10, dtype=torch.float32)
-        weight_dropped_linear(input_tensor)
-        assert all(parameter.is_leaf for parameter in weight_dropped_linear.parameters())
-
-        # Case 4: After reset()
-        weight_dropped_linear.reset()
-        assert all(parameter.is_leaf for parameter in weight_dropped_linear.parameters())
-
-        # Case 5: When in eval mode
+        input_tensor = torch.ones(_in_dim, dtype=torch.float32)
+        target_tensor = torch.zeros(_out_dim, dtype=torch.float32)
+        sgd = torch.optim.SGD(weight_dropped_linear.parameters(), lr=0.01)
+        _assert_sgd_states(sgd, has_grad=False)
+
+        loss_fn = torch.nn.L1Loss()
+        for epoch in range(2):
+            sgd.zero_grad()
+            output_tensor = weight_dropped_linear(input_tensor)
+
+            # Replaced
+            # `assert all(parameter.is_leaf for parameter in weight_dropped_linear.parameters())`
+            # with the following `_assert_sgd_states()`.
+            # Because weight dropped parameters are not deleted after `forward()`.
+            _assert_sgd_states(sgd, has_grad=False if epoch == 0 else True)
+            loss_fn(output_tensor, target_tensor).backward()
+            _assert_sgd_states(sgd, has_grad=True)
+            sgd.step()
+            _assert_sgd_states(sgd, has_grad=True)
+
+        # Case 4: When in eval mode
         weight_dropped_linear.eval()
-        assert all(parameter.is_leaf for parameter in weight_dropped_linear.parameters())
+        # The duplicated-non-leaf weight still exists.
+        pre_eval_parameters = list(weight_dropped_linear.parameters())
+        assert len(pre_eval_parameters) == _n_params + 1
+        assert not all(parameter.is_leaf for parameter in pre_eval_parameters)
+        weight_dropped_linear(input_tensor)
+        # But only the raw weight applies.
+        eval_parameters = list(weight_dropped_linear.parameters())
+        assert len(eval_parameters) == _n_params
```

Comment:
This snippet illustrates my key concern - the length of the parameter list changes depending on whether the module is in eval mode, and whether or not the forward method has been invoked. Presumably if the model state was saved after one of these steps we could run into another issue loading the state dict due to unexpected/missing parameters.

Note: I am not 100% sure whether this issue would occur in practice. My point is that parameters are usually fixed once a module is instantiated, however that does not appear to be the case here, which is difficult to reason about.

Reply:
Hi @rloganiv and @matt-gardner,
Happy new year!
Yes, I'm aware of the side-effect. In my opinion, because the current revision before this PR has another issue of parameter list and model state, so `DropConnect` implementation here hasn't been a pure function, and the trading one side-effect with another seems an acceptable deal.
That being said, I also really want `DropConnect` to be a pure function. Although admittedly I haven't figured out what is an ideal implementation, [the approach of torchnlp](https://github.com/PetrochukM/PyTorch-NLP/blob/master/torchnlp/nn/weight_drop.py) is intriguing. If you guys have other ideal designs in mind, please let me know. I can try to spend some more time on making `DropConnect` pure.